### PR TITLE
[DI] Implement PSR-11

### PR DESCRIPTION
**Pull request by greg0ire**
TODO:

- [x] wait for a stable version of the psr/container package;
- [x] ~~deprecate instanciating ServiceNotFoundException directly, or using any of its methods directly;~~ not relevant anymore
- [x] act on the outcome of https://github.com/php-fig/container/issues/8 (solved in https://github.com/php-fig/container/issues/9) ;
- [x] ~~solve the mandatory NotFoundExceptionInterface on non-existing service
problem (with a flag ?);~~ non-issue, see comments below
- [x] provide meta-package psr/container-implementation if all problems can
be solved.

| Q             | A
| ------------- | ---
| Branch?       | master
| New feature?  | yes
| BC breaks?    | not at the moment
| Tests pass?   | didn't pass before pushing, or even starting to code, will see Travis
| License       | MIT
| Doc PR        | TODO

This PR is a first attempt at implementing PSR-11, or at least trying to get closer to it.
Delegate lookup is optional, and thus not implemented for now.

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/ContainerInterface.php
Comment (greg0ire) [inline src/Symfony/Component/DependencyInjection/ContainerInterface.php]: I chose to have `ContainerInterface` extend `PsrContainerInterface`, so that any interface is marked as psr-compliant. Not sure it is the right thing to do though, because php cannot enforce all the contracts defined in the interface docs (like what exception to throw when). Tell me if you feel I should remove this and add `implements ContainerInterface` in the main implementation.

File: src/Symfony/Component/DependencyInjection/Exception/InvalidArgumentException.php
Comment (greg0ire) [inline src/Symfony/Component/DependencyInjection/Exception/InvalidArgumentException.php]: `EnvNotFound` extends this exception, so no need to add implementation declaration in it.

**Comment by greg0ire** [comment]
About "the mandatory NotFoundExceptionInterface on non-existing service problem", throwing an exception on missing service id is optional in the current implementation while mandatory in the PSR. How I we solve this problem? I was thinking we could deprecate the second argument of `get`, but can we pretend this is a valid implementation is using get with a special argument makes it no longer compliant?

File: src/Symfony/Component/DependencyInjection/composer.json
Comment (Koc) [inline src/Symfony/Component/DependencyInjection/composer.json]: add also provide section https://getcomposer.org/doc/04-schema.md#provide , like in https://github.com/Seldaek/monolog/blob/master/composer.json#L52-L54

Comment (greg0ire) [inline src/Symfony/Component/DependencyInjection/composer.json]: See TODO list ;)

File: src/Symfony/Component/DependencyInjection/ContainerInterface.php
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/ContainerInterface.php]: If it works for PHP, fine.

File: src/Symfony/Component/DependencyInjection/composer.json
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/composer.json]: Still please add it.

**Review by nicolas-grekas — COMMENTED** [review]
There is no issue with throwing : the PSR standardize only one argument, and says to throw. Fortunately that's the default behavior so anyone using the container in the PSR11 way will have it work as expected.
When the 2nd argument is used, we're not in PSR11 anymore, thus we can behave as we want.

File: src/Symfony/Component/DependencyInjection/composer.json
Comment (greg0ire) [inline src/Symfony/Component/DependencyInjection/composer.json]: Fine 👍

**Comment by greg0ire** [comment]
> When the 2nd argument is used, we're not in PSR11 anymore

Makes sense. We can't hurt people that don't know which implementation they are using, because if they don't they will probably not risk calling `get` with two arguments.

File: composer.json
Comment (Soullivaneuh) [inline composer.json]: Just reminder comment: Remove `@dev` annotation when the package will be stable.

**Review by soullivaneuh — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/DependencyInjection/ContainerInterface.php
Comment (keradus) [inline src/Symfony/Component/DependencyInjection/ContainerInterface.php]: that basically means in whole framework you will not use `Psr` version, but `Symfony` version.
Then, one will be able to use concrete class that implements `Symfony` version as parameter for method expecting `Psr` version, but not the other way around. So, if I will have some package that follows `Psr`, I won't be able to use it directly in Symfony project, because everywhere the framework will require `Symfony` version instead

**Comment by mnapoli** [comment]
@greg0ire

> Makes sense. We can't hurt people that don't know which implementation they are using, because if they don't they will probably not risk calling get with two arguments. So it's indeed a non-issue.

:+1: the interface was designed to be compatible with Symfony's default behavior when called with the PSR signature.

**Comment by greg0ire** [comment]
> 👍 the interface was designed to be compatible with Symfony's default behavior when called with the PSR signature.

I did not read the underlying discussions for this spec but there definitely was a "well this is convenient" feeling when implementing it in Symfony 😄

**Comment by dunglas** [comment]
It looks only small changes are required, nice!

**Comment by greg0ire** [comment]
@dunglas yup, that was a breeze :)

**Comment by nicolas-grekas** [comment]
RuntimeException also isn't it?

**Comment by nicolas-grekas** [comment]
Why not make ExceptionInterface extend Psr\Container\ContainerExceptionInterface?

**Comment by greg0ire** [comment]
> Why not make ExceptionInterface extend Psr\Container\ContainerExceptionInterface?

Only exceptions thrown directly by the container are supposed to implement this, so I only applied it to exceptions listed [here](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/DependencyInjection/Container.php#L14)

**Comment by greg0ire** [comment]
> RuntimeException also isn't it?

Can't find it in Container.php, so no.

**Comment by nicolas-grekas** [comment]
In fact, it can, see https://github.com/symfony/symfony/blob/master/src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php#L601

**Comment by nicolas-grekas** [comment]
And https://github.com/symfony/symfony/blob/master/src/Symfony/Component/DependencyInjection/ContainerBuilder.php#L858

**Comment by nicolas-grekas** [comment]
And https://github.com/symfony/symfony/blob/master/src/Symfony/Component/DependencyInjection/ContainerBuilder.php#L374

**Comment by nicolas-grekas** [comment]
I suggest to play simpe here. We don't care to add the interface on all exceptions. This wouldn't hurt.

**Comment by greg0ire** [comment]
> In fact, it can, see https://github.com/symfony/symfony/blob/master/src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php#L601

Yeah but : 

> Exceptions directly thrown by the container 

What does "directly" mean here?

**Comment by nicolas-grekas** [comment]
I don't know if fact, and that's a good question!
But throwing this exception in other cases is not going to make things not-PSR11 compliant. It just would be normalized use case. Fine, nothing broken.
+ I showed you that in fact, via ContainerBuilder, there is a much greater variety of exceptions that can be thrown out there.

**Comment by greg0ire** [comment]
Very true, plus, less code, so… I'll change that. @mnapoli , can you advise about the meaning of "directly" here though?

**Comment by greg0ire** [comment]
There. The diff is tiny now :)

**Comment by greg0ire** [comment]
> This means this is an issue doesn't it ? Should I change this line so that the exception thrown here is wrapped into another?

@nicolas-grekas what about this? Any thoughts?

File: src/Symfony/Component/DependencyInjection/ContainerInterface.php
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/ContainerInterface.php]: @greg0ire without breaking BC, with the benefit if reducing the feature set for the sake of PSR11 compliance? I doubt this is going to happen...

**Comment by nicolas-grekas** [comment]
> Should I change this line so that the exception thrown here is wrapped into another

maybe that's where "directly" applies, meaning "no" to you question.

Why did you remove applying the NotFoundExceptionInterface where applicable?

File: src/Symfony/Component/DependencyInjection/ContainerInterface.php
Comment (greg0ire) [inline src/Symfony/Component/DependencyInjection/ContainerInterface.php]: I'm not talking about reducing the feature set, I'm saying, if we use the container in our internal class A, and we only call `get` and `has` on it, then we should change the type hinting. Not sure how this plays with extension, let me check

Comment (greg0ire) [inline src/Symfony/Component/DependencyInjection/ContainerInterface.php]: Hmm… seems like it would indeed be a BC-break so… only for final classes I guess. https://3v4l.org/I8jO2

Comment (greg0ire) [inline src/Symfony/Component/DependencyInjection/ContainerInterface.php]: Not sure why php complains about it though : every call to test you can make on C, you can make on D, right?

**Comment by greg0ire** [comment]
> maybe that's where "directly" applies, meaning "no" to you question.

indeed :+1: 

>  Why did you remove applying the NotFoundExceptionInterface where applicable?

looks like I was to hasty, sorry.

File: src/Symfony/Component/DependencyInjection/ContainerInterface.php
Comment (jvasseur) [inline src/Symfony/Component/DependencyInjection/ContainerInterface.php]: Because PHP isn't aware of class hierarchy at compile time so it can't allow changing the type declaration.

**Comment by greg0ire** [comment]
> maybe that's where "directly" applies, meaning "no" to you question.

@mnapoli I'll make a PR to clarify that if you confirm

**Comment by greg0ire** [comment]
> looks like I was to hasty, sorry.

@nicolas-grekas I think I got it right this time

File: src/Symfony/Component/DependencyInjection/ContainerInterface.php
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/ContainerInterface.php]: @jvasseur yes it does
the issue is that `$foo instanceof A` implies that calling `$foo->test($anotherInstanceOfA)` is allowed.
Since `$anInstanceofB instanceof A`, calling `$anInstanceofB->test($anotherInstanceOfAButNotB)` must thus be allowed.

Comment (greg0ire) [inline src/Symfony/Component/DependencyInjection/ContainerInterface.php]: So… I found a bug in php?

Comment (jvasseur) [inline src/Symfony/Component/DependencyInjection/ContainerInterface.php]: It's a well know limitation of the PHP engine : https://bugs.php.net/bug.php?id=69612#1431343525

Comment (jvasseur) [inline src/Symfony/Component/DependencyInjection/ContainerInterface.php]: But your example is wrong anyway, functions parameters could be contravariant (they are invariant in PHP) meaning they can accept a wider type. In your example they are covariant instead.

Comment (greg0ire) [inline src/Symfony/Component/DependencyInjection/ContainerInterface.php]: Oh you're right, my logic was wrong indeed! Which means this would indeed be a BC-break, sorry I got confused.

**Comment by moufmouf** [comment]
Hey @greg0ire ,

Looking at the current implementation you are proposing, regarding exceptions, there is a slight hickup.

The `ServiceNotFoundException` from Symfony is used both for saying that the user requested a given service and that a dependency is missing (i.e. one service is performing a lookup on another service that does not exists).

According to PSR-11, we should have 2 different exceptions for this (because only the first case should implement the `ServiceNotFoundExceptionInterface`).

The problem is hopefully easy enough to solve.
I fixed it in a fork from your fork (inception!), here: https://github.com/moufmouf/symfony/commit/9960c36ae6f42817c34e644df369d6661a46d223

Please merge that change if you find it appropriate.

To solve this issue and keep backward compatibility:

- I created 2 new exceptions: `RequestedServiceNotFoundException` and `MissingDependencyException`.
- Both exceptions extend the `ServiceNotFoundException` (this way, existing code catching this exception still works)
- Only the `RequestedServiceNotFoundException` implements the `ServiceNotFoundExceptionInterface`

@nicolas-grekas I've seen your comments on the PHP-FIG mailing list and it's great to have Symfony feedback. We value your opinion and I have a very concrete question: do you think we should standardize a `MissingDependencyExceptionInterface`? Do you think PSR-11 would be a better standard if we do so?

// cc @mnapoli

**Comment by nicolas-grekas** [comment]
At first we think that implementing PSR-11 is trivial - then the exception handling part comes in and the diff becomes a lot less trivial.
Devil is in the detail (same lesson learned while implementing PSR-16 btw, nothing specific to PSR-11, nor to any PSR really).

> do you think we should standardize a MissingDependencyExceptionInterface? Do you think PSR-11 would be a better standard if we do so?

I do :)

but there are remaining concerns in the recent comments.
Dunno how to be more constructive here sorry... Fortunately I don't vote on the FIG so it doesn't matter for you if I miss the point :)

**Comment by nicolas-grekas** [comment]
Maybe this, to be constructive: #20658
I think a PSR with what's currently in PSR-11 + a way for classes/objects to declare a set of services could maybe be worth it.

**Comment by greg0ire** [comment]
@moufmouf @nicolas-grekas I think there might be a contradiction here:

> But throwing this exception in other cases is not going to make things not-PSR11 compliant. It just wouldn't be a normalized use case. Fine, nothing broken.

> because only the first case should implement the ServiceNotFoundExceptionInterface

So which is it? Can exceptions that are not supposed to implement a psr interface implement it anyway or not?

**Comment by greg0ire** [comment]
After more thoughts @moufmouf @nicolas-grekas I think you're both right, I just couldn't understand why, but it might be because `ContainerInterface` is just a marker interface, while `ServiceNotFoundExceptionInterface` conveys meaning. Would that be why you did not add the suffix to the latter exception [at first](https://github.com/php-fig/container/pull/9/files), but did add it to the former?

**Comment by greg0ire** [comment]
>  I fixed it in a fork from your fork (inception!), here: moufmouf@9960c36

I went all the way through with your inception concept and created a meta-PR : https://github.com/greg0ire/symfony/pull/1/files (will rebase once merged though)

**Comment by greg0ire** [comment]
Added @moufmouf 's commit to my PR

File: src/Symfony/Component/DependencyInjection/composer.json
Comment (GuilhemN) [inline src/Symfony/Component/DependencyInjection/composer.json]: Looks like should also be added to the root `composer.json`.

Comment (greg0ire) [inline src/Symfony/Component/DependencyInjection/composer.json]: Uuuuh… it's already there, isn't it?

Comment (GuilhemN) [inline src/Symfony/Component/DependencyInjection/composer.json]: Indeed... I missed it 😳

Comment (greg0ire) [inline src/Symfony/Component/DependencyInjection/composer.json]: Thanks for reviewing anyway :)

File: src/Symfony/Component/DependencyInjection/Exception/RequestedServiceNotFoundException.php
Comment (keradus) [inline src/Symfony/Component/DependencyInjection/Exception/RequestedServiceNotFoundException.php]: `===` btw ;)

File: src/Symfony/Component/DependencyInjection/Exception/MissingDependencyException.php
Comment (keradus) [inline src/Symfony/Component/DependencyInjection/Exception/MissingDependencyException.php]: could we make them as final ?

File: src/Symfony/Component/DependencyInjection/Exception/RequestedServiceNotFoundException.php
Comment (greg0ire) [inline src/Symfony/Component/DependencyInjection/Exception/RequestedServiceNotFoundException.php]: `count` only returns integers, but ok…

**Comment by greg0ire** [comment]
@keradus : made exceptions introduced by @moufmouf final.

File: src/Symfony/Component/DependencyInjection/Exception/ServiceNotFoundException.php
Comment (unkind) [inline src/Symfony/Component/DependencyInjection/Exception/ServiceNotFoundException.php]: BC break?

Comment (greg0ire) [inline src/Symfony/Component/DependencyInjection/Exception/ServiceNotFoundException.php]: Classes that extend this exception will no longer get this method, so technically yeah. @moufmouf @nicolas-grekas , should I restore it, for BC's sake, or do we deem that "acceptable BC-break" ?

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Exception/ServiceNotFoundException.php]: changing a public interface is not an acceptable BC break - the class cannot be abstract for the same reason btw

Comment (greg0ire) [inline src/Symfony/Component/DependencyInjection/Exception/ServiceNotFoundException.php]: Ok let's fix this.

Comment (greg0ire) [inline src/Symfony/Component/DependencyInjection/Exception/ServiceNotFoundException.php]: What about the constructor? Removing it will not cause anything to crash, but then if it doesn't set `sourceId`, it's a BC-break, so in fact I'm starting to think I should restore the entirety of this class to its former self.

Comment (moufmouf) [inline src/Symfony/Component/DependencyInjection/Exception/ServiceNotFoundException.php]: Good question. Are constructors of exceptions supposed to be part of public API? Is there a good reason for anyone to instantiate a ServiceNotFoundException except the symfony container itself?

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Exception/ServiceNotFoundException.php]: @moufmouf yes to both questions At least that's the default answer, and if one thinks otherwise, one must demonstrates why at the technical level. Here, I doubt this is possible. So: yes and yes.

Comment (unkind) [inline src/Symfony/Component/DependencyInjection/Exception/ServiceNotFoundException.php]: Given `ContainerInterface::get()` declares this exception, one can decorate `ContainerInterface` and instantiate it.

**Comment by greg0ire** [comment]
> given the recent discussion on the FIG about it, maybe you should just wait
> because it might not be necessary to do this change at all in the end 

— @nicolas-grekas , [on the symfony dev slack channel](https://symfony-devs.slack.com/archives/general/p1484559452004865)

I'm putting this on hold until said discussion is resolved

**Comment by nicolas-grekas** [comment]
Now that https://github.com/php-fig/fig-standards/pull/869 has been merged, I think we can remove the second commit here, isn't it?

**Comment by greg0ire** [comment]
> Now that php-fig/fig-standards#869 has been merged

TL;DR `NotFoundExceptionInterface` is useless

**Comment by greg0ire** [comment]
I also should rework the first commit so that it does not reference this interface anymore, shouldn't I?

**Comment by nicolas-grekas** [comment]
Rebase needed + missing autowiring-type for `service_container` in `src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml`

**Comment by greg0ire** [comment]
@nicolas-grekas fixed

File: composer.json
Comment (nicolas-grekas) [inline composer.json]: alpha order

Comment (greg0ire) [inline composer.json]: I must be tired :P

Comment (greg0ire) [inline composer.json]: fixed

File: composer.json
Comment (fabpot) [inline composer.json]: Can we remove it now so that this PR is ready to be merged?

Comment (greg0ire) [inline composer.json]: fixed

**Comment by greg0ire** [comment]
Travis passes although there is no stable release yet… I don't get it.

**Comment by nicolas-grekas** [comment]
another rebase please? should be the last :)

**Comment by greg0ire** [comment]
Oh wow bye bye autowiring huh? I noticed the service was no longer synthetic too, is that normal? Not sure I got it right…

**Comment by nicolas-grekas** [comment]
You got it right. See #21494 for more info about this change.
The `service_container` is already declared implicitly, it was here only to add autowiring-types, but these are gone now.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml]: Psr

Comment (greg0ire) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml]: Sorry, I retried with auto-completion this time…

**Comment by chalasr** [comment]
Needs review?

**Comment by greg0ire** [comment]
> Needs review?

The more eyeballs the better, not sure you are going to find much to say now though.

I think the `Needs work` label could maybe be removed, this is more `blocked`, because we are waiting for a release of the psr dependency.

**Review by nicolas-grekas — APPROVED** [review]
:+1:

**Comment by dunglas** [comment]
👍

**Review by dunglas — APPROVED** [review]

**Review by javiereguiluz — APPROVED** [review]
👍

**Comment by fabpot** [comment]
Thank you @greg0ire.

**Comment by greg0ire** [comment]
Thanks for merging!

**Comment by allflame** [comment]
Can you please consider BC breaks not only with respect to "implementation" but also to "declaration":
This is how it was before:
```
interface I1 {} 
interface I2 {}
class C1 implements I1, I2 {}
```
And that's what happened:
```
interface I1 extends I2 {};
Class C1 implements I1, I2 -> Fatal Error
```
Your interfaces are your burden as much as your implementation.

**Comment by greg0ire** [comment]
@allflame: can't reproduce https://3v4l.org/gNuOY

**Comment by jvasseur** [comment]
@greg0ire you have to change the order of the implemented interfaces : https://3v4l.org/032Sk

It's a know PHP bug.

**Comment by greg0ire** [comment]
Oh ok... well sorry, but I didn't know that bug.

**Comment by xabbuh** [comment]
Which class should be affected by that? This PR did not change any class to implement two interfaces, did it?

**Comment by greg0ire** [comment]
@xabbuh classes outside Symfony that happen to implement `ContainerInterface`and `PsrContainerInterface` (not sure what order is buggy but one of them is)

**Comment by xabbuh** [comment]
Hm yeah, nothing we can account for here. This needs to be changed in those projects or fixed in PHP.

**Comment by allflame** [comment]
@xabbuh this "bug" is like 5 years old and since it just a consequence of OOP implementation in PHP it won't likely be "fixed" for another 5 years. My message here is the following: changes in "contracts/definitions" (that is interfaces, inheritance, function declaration etc) have way more potential to break things than changes in the "implementation" (e.g. actual code)
And one thing that you didn't know about PHP behavior and made that change - that's fine, but if you are aiming for BC in minor release - you cannot make statements like that, knowing this will break things (for whatever reasons)

**Comment by stof** [comment]
Well, this means that we would have to avoid providing a PSR-11 compliant container because someone might have already built an external project implementing both PSR-11 and our interface ? This means we would never be able to implement PSR on existing code outside major versions.

Btw, our BC promise explicitly mentions that we *don't* ensure BC if you add new methods in a child class (as this would then forbid us to add any method in our existing classes due to potential signature mismatches for the child class using the name already). Adding extra interfaces on the class generally involve adding such method (PSR-11 actually may not need it though)

**Comment by allflame** [comment]
@stof Yes, this is what software versioning, BC and PHP imply.
The code snippet I've provided has no child classes and I don't understand why you mentioned it here. The real cause was adding extra interfaces to interface which introduced incompatibility. I'll let you figure out why this differs so much with your classes example.

**Comment by stof** [comment]
Well, are you actually implementing the Symfony ContainerInterface from scratch without using our base classes ?

**Comment by allflame** [comment]
No, of course not.
My personal believe is that most Symfony code is very good to use as is, that's why I'm using it in other frameworks trying to substitute their own realization.
Original piece of code was like this:
https://github.com/allflame/vain-phalcon/blob/v0.4.104/src/Di/Symfony/SymfonyContainerAdapter.php#L26
This has to be done, because part of the framework depends on PhalconInterface, Symfony on its own depends on SymfonyInterface and the whole application is container-agnostic (well, up to PSR-11), hence PsrContainerInterface.
And that's when this commit backfired.

**Pull request merged**
